File: java/headFirstDesignPatterns/chap3/starbuzz/src/br/com/starbuzz/drinks/Soy.java
package br.com.starbuzz.drinks;

public class Soy extends CondimentDecorator {

    Beverage beverage;
    
    public Soy(Beverage beverage) {
        this.beverage = beverage;
    }
    
    @Override
    public String getDescription() {
        return this.beverage.getDescription()+", Soy";
    }

    @Override
    public double cost() {
        return .15 + this.beverage.cost();
    }

}

File: c_cpp/headFirst_designPatterns/chap3/starbuzz/src/options/Soy.cpp
/*
 * Soy.cpp
 *
 *  Created on: 24/01/2015
 *      Author: jcarlos
 */

#include "Soy.hpp"

namespace starbuzz {
namespace options {

Soy::Soy(Beverage* beverage) {
    this->beverage = beverage;
}

Soy::~Soy() {
    delete beverage;
}

const char* Soy::getDescription(){
    string str = this->beverage->getDescription();
    str += ", Soy";
    return str.c_str();
}

double Soy::getCost(){
    return this->beverage->getCost() + 0.15;
}

} /* namespace options */
} /* namespace starbuzz */
